core/include/vecmem/containers/details/jagged_device_vector_iterator.hpp

```diff
@@ -16,190 +16,185 @@
-      reference operator*() const;
-      /// Use the iterator as a pointer
-      VECMEM_HOST_AND_DEVICE
-      pointer operator->() const;
-      /// Return the value at a specific offset
-      VECMEM_HOST_AND_DEVICE
-      reference operator[]( difference_type n ) const;
-
-      /// @}
-
-      /// @name Iterator updating operators
-      /// @{
-
-      /// Decrement the underlying iterator (with '++' as a prefix)
-      VECMEM_HOST_AND_DEVICE
-      jagged_device_vector_iterator& operator++();
-      /// Decrement the underlying iterator (wuth '++' as a postfix)
-      VECMEM_HOST_AND_DEVICE
-      jagged_device_vector_iterator operator++( int );
-
-      /// Increment the underlying iterator (with '--' as a prefix)
-      VECMEM_HOST_AND_DEVICE
-      jagged_device_vector_iterator& operator--();
-      /// Increment the underlying iterator (with '--' as a postfix)
-      VECMEM_HOST_AND_DEVICE
-      jagged_device_vector_iterator operator--( int );
-
-      /// Decrement the underlying iterator by a specific value
-      VECMEM_HOST_AND_DEVICE
-      jagged_device_vector_iterator operator+( difference_type n ) const;
-      /// Decrement the underlying iterator by a specific value
-      VECMEM_HOST_AND_DEVICE
-      jagged_device_vector_iterator& operator+=( difference_type n );
-
-      /// Increment the underlying iterator by a specific value
-      VECMEM_HOST_AND_DEVICE
-      jagged_device_vector_iterator operator-( difference_type n ) const;
-      /// Increment the underlying iterator by a specific value
-      VECMEM_HOST_AND_DEVICE
-      jagged_device_vector_iterator& operator-=( difference_type n );
-
-      /// @}
-
-      /// @name Comparison operators
-      /// @{
-
-      /// Check for the equality of two iterators
-      VECMEM_HOST_AND_DEVICE
-      bool operator==( const jagged_device_vector_iterator& other ) const;
-      /// Check for the inequality of two iterators
-      VECMEM_HOST_AND_DEVICE
-      bool operator!=( const jagged_device_vector_iterator& other ) const;
-
-      /// @}
+    /// @name Types describing the underlying data
+    /// @{
+
+    /// Type of the data object that we have an array of
+    typedef data::vector_view<TYPE> data_type;
+    /// Pointer to the data object
+    typedef data_type* data_pointer;
+
+    /// @}
+
+    /// @name Type definitions, mimicking STL iterators
+    /// @{
+
+    /// Value type being (virtually) iterated on
+    typedef device_vector<TYPE> value_type;
+    /// (Pointer) Difference type
+    typedef std::ptrdiff_t difference_type;
+    /// "Reference" type to the underlying (virtual) value
+    typedef value_type reference;
+
+    /// Helper class for returning "pointer-like" objects from the iterator
+    ///
+    /// Since the iterator returned everything by value as temporary objects,
+    /// in order to provide a proper return type for its @c operator->, this
+    /// custom type needs to be used.
+    ///
+    class pointer {
+       public:
```

Comment:
This is a bit difficult to set up so that a single configuration would result in nice looking code in all circumstances, but I think we should use `EmptyLineBeforeAccessModifier: Leave`...

I really don't like the Clang default of showing the first access modifier right against the class declaration, while leaving the other modifiers afterwards with nice empty lines before them. For me personally it makes it more difficult to quickly identify the different access blocks inside of a class...

But I'm open to a debate on this one.

Reply:
I've never had readability issues with this. This would require us going to clang-format12, and I'm not sure it's that important.

Reply:
`KeepEmptyLinesAtTheStartOfBlocks` helps with this somehow, because you can manually do it.